### ISSUE TITLE
Replay mod 

### ISSUE THREAD

**Issue by CarbonNeuron** [ORelio/Minecraft-Console-Client#1007]
**Prerequisites**
 - [x] I have read and understood the [user manual](https://github.com/ORelio/Minecraft-Console-Client/tree/master/MinecraftClient/config)
 - [x] I made sure I am running the latest [development build](https://ci.appveyor.com/project/ORelio/minecraft-console-client/build/artifacts)
 - [x] I tried to [look for similar feature requests](https://github.com/ORelio/Minecraft-Console-Client/issues?q=is%3Aissue) before opening a new one

**Feature description**
A feature request should contain the following items:
* A concise description of your problem (I'm always frustrated when [...])
* A clear description of what you want to happen (I'd like the app to do this to solve my problem [...])
* Any alternative solutions or features you've considered.
* If applicable, add screenshots to help explain your request

So basically replay mod is a mod for Minecraft that records your game, by recording the packets. Then you can use the replay viewer which will replay your packets. And you can export it as an mp4 file, this is useful for factions. As no handling needs to be done for graphics, you just need to write down the packets, I feel this may be something that wouldn't require that much work. I've tried to integrate but I've been unsuccessful. Hopefully someone else could take a jab at it

**Comment by ReinforceZwei** [comment]
Replay mod uses a particular file format `.mcpr` for saving the packets data and seems like it is encoded, and we need to figure out the file format before we can integrate them. 
Referring to their [forums](https://www.replaymod.com/forum/thread/307), currently there is no way to edit the `.mcpr` file.
As long as we figured out the file format, we could start implement this feature.

But I have no idea how to figure it out :cry:

**Comment by TheSnoozer** [comment]
`.mcpr` files used to be just `zip` files with at least a `metaData.json` and `recording.tmcpr` file inside.

The `.tmcpr` file can be read using a DataInputStream in Java.
The following pattern repeats until EOF:

1) an Integer which represents the packet's timestamp (in milliseconds from Replay start)
2) an Integer which tells you how many bytes the packet consists of
3) a byte array which represents the packet

More details about the format:
- https://www.replaymod.com/forum/thread/609#post3335
- https://github.com/ReplayMod/ReplayMod/issues/49

**Comment by ReinforceZwei** [comment]
Base on the information provided by @TheSnoozer , I was able to read and "parse" the `.tmcpr` file.
![image](https://user-images.githubusercontent.com/39955851/90848979-c9e93d80-e3a0-11ea-8411-ab163c5ae859.png)
Next I am going to try writing the `.tmcpr` file using MCC and let Replay mod to read it.

**Comment by Pokechu22** [comment]
[`System.IO.Compression.ZipFile`](https://docs.microsoft.com/en-us/dotnet/api/system.io.compression.zipfile?view=netcore-3.1)?  Or is that not part of the framework you have access to?

**Comment by ReinforceZwei** [comment]
@Pokechu22 That requires.net framework 4.5+ but MCC still on 4 so we can't use that.

**Comment by ORelio** [comment]
Congrats, @ReinforceZwei!

> Do replay mod translate those movement packet from C2S to S2C then save them?

After a quick look to the repository I read that the replay mod itself adds some packets although I can't find where I read this. However, I found this event handler that, indeed, seems to generate fake packets for recoding the player's location: [RecordingEventHandler.java#L144](https://github.com/ReplayMod/ReplayMod/blob/2ab92f554fd51dd71293aab5baf1d79210f41943/src/main/java/com/replaymod/recording/handler/RecordingEventHandler.java#L144).

> Do you know any good library or solution for making zip file? _(no compression is okay, replay mod can read it)_

MCC uses parts of code from [DotNetZip](https://archive.codeplex.com/?p=dotnetzip) in the [Compression](https://github.com/ORelio/Minecraft-Console-Client/tree/master/MinecraftClient/Protocol/Handlers/Compression) namespace for handling Zlib compression of packets. I also used DotNetZip for another .NET4 project ([SoundManager](https://github.com/ORelio/Sound-Manager)), and creating a Zip file is straightforward:
```C#
string filePath = //...
using (ZipFile zip = new ZipFile())
{
    zip.AddFile(filePath, "");
    //zip.AddFile(filePath2, "");
    //...
    zip.Save(zipfile);
}
```
You might be able to add more files to the Compression namespace or add a ZipFile namespace.

**Comment by ReinforceZwei** [comment]
I imported the dotNetZip library into MCC and I can use the ZipFile class now. But whatever how I use the ZipFile class, it won't add the binary file `.tmcpr` into the zip file. I have to do something like this in order to create a working zip file:
![image](https://user-images.githubusercontent.com/39955851/91016215-02964a80-e61f-11ea-92ae-05052c3ac182.png)

Now MCC can generate `.mcpr` file and can be read directly by replay mod. :tada:

**Comment by CarbonNeuron** [comment]
How is performance on this?

**Comment by ReinforceZwei** [comment]
@CarbonNeuron Not sure what performance you are referencing.

Basic function are done except the client player is not showing up in the replay. Don't know what kinds packet need to be added into the replay for client player to shown properly. Also it is a bit hard to translate packet from outbound to inbound because the packet parsing design of MCC is single way. _I can't even get the outgoing packet ID by type using the existing code._

**Comment by ORelio** [comment]
The packets for player position are linked above in RecordingEventHandler.java, it's server to client player teleport and player movement and look. But yes mappings in MCC are one-way (id to type, not type to id) so basically you'd need to reimplement packet generation for these packets in order to add them in the replay. Protocol18PacketTypes.cs is like palette handling but with a bunch of switches instead of proper dictionary because palette handling did not exist in MCC at the time and it was easier to implement. You may add an additional GetPacketIncomingID with only the packet IDs you need.

**Comment by ReinforceZwei** [comment]
_Handling packet types and IDs are more comfortable now_
I have basic concept of which packet should be translated for player movement, even though MCC client player aren't likely will be moved I guess :P
But I have no success on spawning the client player, so I couldn't do the movement part.
![image](https://user-images.githubusercontent.com/39955851/91424581-915bcf00-e88c-11ea-9a18-fae7cc7f9a52.png)

**Comment by ReinforceZwei** [comment]
I decided to give-up the client player. Now I need some good idea on how to enable or disable this Replay handler through ini setting since it injected couple methods call directly into the Protocol18Handler. 

I have thought of two solutions:
- Set a boolean field in Replay handler to enable or disable the functions
- Use ChatBot and add a new `OnPacket` event to it. Then we don't need to inject the Protocol18Handler

@ORelio Which solution do you think is better? Or do you have a better solution?

**Comment by ORelio** [comment]
`OnNetworkPacket` seems a good idea, this makes easy to implement packet recording, and custom chatbots could benefit from raw packet access. The only catch about this is that it requires to create a copy of the packet in memory, slowing down the client when no chatbots are actively processing raw network packets. ChatBots willing to record packets should enable packet recording first using an additional setting or ChatBot API method. So yeah, I'd suggest to use both a ChatBot event _and_ a boolean field to enable packet capture for use in ChatBots. In `Protocol18Handler`, you'd just need to check if the boolean is set (for instance in `Settings.cs` or in `McClient.cs` through a getter method), and if so, make a copy of the packet in an array or list, and pass it to McClient as a regular event.

**Comment by ReinforceZwei** [comment]
Such a great idea of combining them! I will take this. Thanks!
